[aes/dv] Transition to V1 stage

vogelpi opened the pull request:
With Rasmus' PR #1319 having landed, we reached a big milestone towards the verification of AES. This should be reflected in the [hardware dashboard](https://docs.opentitan.org/doc/project/hw_dashboard/). 

For the AES module to reach the next V stage, we need make sure all points in the checklist are indeed fulfilled. I quickly went through [our checklist](https://docs.opentitan.org/doc/project/checklist/index.html#v1) and started to update the hjson.

@rasmus-madsen it would be good if you could also have a look, update the hjson and maybe plan your next steps accordingly. We can also have a call to discuss this.

UPDATE: I think we are ready for the review.

rasmus-madsen commented on hw/ip/aes/doc/checklist.md:
```diff
@@ -104,21 +104,21 @@ Review        | Signoff date            | Not Started |
-Tool Setup    | [ALT_TOOL_SETUP][]                    | Not Started |
-Regression    | [SANITY_REGRESSION_SETUP][]           | Not Started |
-Regression    | [NIGHTLY_REGRESSION_SETUP][]          | Not Started |
+Documentation | [TESTPLAN_COMPLETED][]                | Not Started | Not sure if the testplan is complete yet.
```
not completed but definitely started

sriyerg replied:
For V1 qualification, testplan and DV plan needs to be reviewed by designers and peers. I dont think we have had that done yet.

sriyerg replied:
https://docs.opentitan.org/doc/project/hw_stages/#hardware-verification-stages

rasmus-madsen commented on hw/ip/aes/doc/checklist.md:
```diff
@@ -104,21 +104,21 @@ Review        | Signoff date            | Not Started |
-Regression    | [NIGHTLY_REGRESSION_SETUP][]          | Not Started |
+Documentation | [TESTPLAN_COMPLETED][]                | Not Started | Not sure if the testplan is complete yet.
+Testbench     | [TB_TOP_CREATED][]                    | Done        |
+Testbench     | [PRELIMINARY_ASSERTION_CHECKS_ADDED][]| Not Started | Not sure
```
no assertion checks yet

rasmus-madsen commented on hw/ip/aes/doc/checklist.md:
```diff
@@ -104,21 +104,21 @@ Review        | Signoff date            | Not Started |
+Testbench     | [RAL_MODEL_GEN_AUTOMATED][]           | Done        |
+Testbench     | [TB_GEN_AUTOMATED][]                  | N/A         |
+Tests         | [SANITY_TEST_PASSING][]               | Done        |
+Tests         | [CSR_MEM_TEST_SUITE_PASSING][]        | Not Started | Not sure
```
we dont have any memories, but all CSR tests are passing

vogelpi replied:
Alright, thanks for your feedback. I'll update that one.

sriyerg commented:
When in doubt, please refer to the checklist definitions for clarity:
https://docs.opentitan.org/doc/project/checklist/

I think there is still a few items TBD before we can call V1 (testplan, dv plan review).

sriyerg commented:
I ran an AES regression at HEAD - looks like none of the CSR tests are passing:

```console
$ ../../../../util/dvsim.py aes_sim_cfg.hjson -i V1 --map-full-testplan -v n

# AES Regression Results
  Run on Thursday January 23 2020 10:22:31AM 

## Test Results
|  Milestone  |        Name        |                    Tests |  Results  |
|:-----------:|:------------------:|-------------------------:|:---------:|
|     V1      |      wake_up       |               aes_wakeup |  100/100  |
|     V1      |       sanity       |               aes_sanity |  100/100  |
|     V1      |        csr         |         aes_csr_hw_reset |   0/40    |
|             |                    |               aes_csr_rw |   0/40    |
|             |                    |         aes_csr_bit_bash |   0/40    |
|             |                    |         aes_csr_aliasing |   0/40    |
|     V1      |                    |                    TOTAL |  200/360  |
|     V2      |     algorithm      |               aes_sanity |  100/100  |
|             |                    |               aes_stress |    0/1    |
|     V2      |      key_size      |               aes_stress |    0/1    |
|     V2      |    failure_test    |               aes_stress |    0/1    |
|     V2      |   reset_recovery   |               aes_stress |    0/1    |
|     V2      |  deinitialization  |               aes_stress |    0/1    |
|     V2      |  oob_addr_access   |            aes_tl_errors |    0/1    |
|     V2      |   illegal_access   |            aes_tl_errors |    0/1    |
|     V2      | outstanding_access |         aes_csr_hw_reset |   0/40    |
|             |                    |               aes_csr_rw |   0/40    |
|             |                    |         aes_csr_aliasing |   0/40    |
|             |                    | aes_same_csr_outstanding |    0/1    |
|     V2      |   partial_access   |         aes_csr_hw_reset |   0/40    |
|             |                    |               aes_csr_rw |   0/40    |
|             |                    |         aes_csr_aliasing |   0/40    |
|     V2      |                    |                    TOTAL |  100/348  |
|             |                    |                    TOTAL |  300/708  |
```

rasmus-madsen commented:
are they run in the regression?
maybe we should offline @sriyerg  I might be missing some info here.

rasmus-madsen commented:
@vogelpi maybe we change this to target 0.5?
and we should discuss a roadmap to 1.0
it might make more sense to target that directly rather than adding more features/modes for now?

sriyerg commented:
> are they run in the regression?
> maybe we should offline @sriyerg I might be missing some info here.

I setup a meeting in 30 minutes - perhaps you can make it.

vogelpi commented:
I updated the checklist and rebased the branch on top of master. @rasmus-madsen will do a PR for the DV documentation, the last missing piece before starting the review for V1.

vogelpi commented:
I after @rasmus-madsen 's DV documentation PR has landed, I think we are ready to kick off the review. I guess this will involve @sriyerg as DV lead and @sjgitty as design lead. Please let me know if you need anything from my/our side.

sjgitty commented:
Sounds good @vogelpi we can do that. Transitions to D1 and/or V1 shouldn't
take too long to review

On Tue, Mar 3, 2020 at 10:06 AM Pirmin Vogel <notifications@github.com>
wrote:

> I after @rasmus-madsen <https://github.com/rasmus-madsen> 's DV
> documentation PR has landed, I think we are ready to kick off the review. I
> guess this will involve @sriyerg <https://github.com/sriyerg> as DV lead
> and @sjgitty <https://github.com/sjgitty> as design lead. Please let me
> know if you need anything from my/our side.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/lowRISC/opentitan/pull/1401?email_source=notifications&email_token=AJZQKVPEXT6DB6CXZZ37DZ3RFVBKJA5CNFSM4KKTQUN2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOENURBKY#issuecomment-594088107>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJZQKVMGE4P66EPRRRQBZELRFVBKJANCNFSM4KKTQUNQ>
> .
>

rasmus-madsen commented:
@sriyerg @weicaiyang  is there anything needed from me to move forward with the review?
we would like this merged before adding new RTL / DV features

weicaiyang commented:
> @sriyerg @weicaiyang is there anything needed from me to move forward with the review?
> we would like this merged before adding new RTL / DV features

Looks like aes_sanity is still failing. Could you fix it as one V1 item is SANITY_TEST_PASSING?
https://reports.opentitan.org/hw/ip/aes/dv/latest/results.html

rasmus-madsen commented:
@weicaiyang this issue is tracked in https://github.com/lowRISC/opentitan/issues/1696
so far it does not seem to be a testcase issue.
I think we can go ahead with the review so we are ready to merge once #1696 is resolved!

sriyerg commented on hw/ip/aes/data/aes.prj.hjson:
```diff
@@ -7,5 +7,5 @@
     life_stage:         "L1",
     design_stage:       "D2",
-    verification_stage: "V0",
+    verification_stage: "V1",
```
Would it be okay to hold off on merging this PR, since there are V1 items that are still "in progress" and "not started"?

vogelpi replied:
Actually, the this is the idea of this PR. We need to trigger the review. We would like to know what is missing to push this forward.

vogelpi replied:
Sorry @sriyerg I did not realize that the V1 review is separate process from the testplan review. We would like to move on, if possible let's do this in one shot.

vogelpi commented on hw/ip/aes/doc/checklist.md:
```diff
@@ -104,21 +104,21 @@ Review        | Signoff date            | Not Started |
-Integration   | [PRE_VERIFIED_SUB_MODULES_V1][]       | Not Started |
-Review        | [DESIGN_SPEC_REVIEWED][]              | Not Started |
-Review        | [DV_PLAN_TESTPLAN_REVIEWED][]         | Not Started |
+Documentation | [TESTPLAN_COMPLETED][]                | Done        |
```
Thanks Sri, i aligned the AES checklist with the one from UART and the signoff checklist in the documentation. I will create a PR to also re-align the template.

sriyerg replied:
Thanks @vogelpi!

vogelpi commented:
I re-aligned the checklist with the UART one (and did a rebase).

There are one more item I am not sure about:
- CSR_CHECK_GEN_AUTOMATED: https://docs.opentitan.org/doc/project/checklist/#v1
@rasmus-madsen could you please check if that is indeed fulfilled?

As for the other items:
- DV_PLAN_TESTPLAN_REVIEWED: This is what this PR is about. We would like to know if you see something missing.
- STD_TEST_CATEGORIES_PLANNED: This is maybe something we could start already @rasmus-madsen ? I think we discussed many ideas a while back. I think it is time to come up with a plan for those.
- V2_CHECKLIST_SCOPED: This goes hand in hand with the previous item. Did you already start this @rasmus-madsen ?

The last two items are independent of the test plan review.
But all item are needed to start the actual V1 review.

rasmus-madsen commented:
> I re-aligned the checklist with the UART one (and did a rebase).
> 
> There are one more item I am not sure about:
> 
> * CSR_CHECK_GEN_AUTOMATED: https://docs.opentitan.org/doc/project/checklist/#v1
>   @rasmus-madsen could you please check if that is indeed fulfilled?
 
I assume this is done by including `/dv/tools/common_tests.mk` @sriyerg ?
if so yes then this is done

> 
> As for the other items:
> 
> * DV_PLAN_TESTPLAN_REVIEWED: This is what this PR is about. We would like to know if you see something missing.
> * STD_TEST_CATEGORIES_PLANNED: This is maybe something we could start already @rasmus-madsen ? I think we discussed many ideas a while back. I think it is time to come up with a plan for those.
@vogelpi I think for Security and error we do need to come make a formalized plan.
the next two I am not sure what is needed from a DV perspective
- power  not sure there is much to do about power in terms of DV?
- Performance @sriyerg I need some elaboration here - performance in terms of simulation time or in terms of algorithm performance? 

for debug and stress I think we already have good handle on these - maybe I don't fully understand the word planned here.
I can add going forward plans to the DV guide but then the state of the document does not match what is in the code. and I don't think it makes sense to create a new document stating what will go into the real document later! 

> * V2_CHECKLIST_SCOPED: This goes hand in hand with the previous item. Did you already start this @rasmus-madsen ?
Yes this is scoped. in terms of man hours not actual time.. 

> 
> The last two items are independent of the test plan review.
> But all item are needed to start the actual V1 review.

sriyerg commented:
> > I re-aligned the checklist with the UART one (and did a rebase).
> > There are one more item I am not sure about:
> > 
> > * CSR_CHECK_GEN_AUTOMATED: https://docs.opentitan.org/doc/project/checklist/#v1
> >   @rasmus-madsen could you please check if that is indeed fulfilled?
> 
> I assume this is done by including `/dv/tools/common_tests.mk` @sriyerg ?
> if so yes then this is done
> 

Yes this can be marked off as done. CSR tests on AES are running in the nightly regressions and are passing: 
https://reports.opentitan.org/hw/ip/aes/dv/latest/results.html

> > As for the other items:
> > 
> > * DV_PLAN_TESTPLAN_REVIEWED: This is what this PR is about. We would like to know if you see something missing.

I was thinking more on the lines of blocking 30 mins of the next Si meeting (or setup an ad-hoc meeting) and presenting the testplan (what all things are being tested) and DV plan (how are those things tested) for a wider audience. It would be hard to do an "offline" review of the PR with all the stakeholders involved. 

> > * STD_TEST_CATEGORIES_PLANNED: This is maybe something we could start already @rasmus-madsen ? I think we discussed many ideas a while back. I think it is time to come up with a plan for those.
> >   @vogelpi I think for Security and error we do need to come make a formalized plan.
> >   the next two I am not sure what is needed from a DV perspective
> 
> * power  not sure there is much to do about power in terms of DV?
> * Performance @sriyerg I need some elaboration here - performance in terms of simulation time or in terms of algorithm performance?

Whatever the design spec mandates. Examples include - AES completing within X number of cycles, meeting specific bandwidth / latency requirements, etc. 
I added this category because it is easy to overlook these categories of tests. Note that not all may be applicable - once that are not can be marked as NA (in the notes section). 

The V milestone checklists are oriented towards achieving a well verified design. While I do recommended the DV engineers to profile the simulations from time to time and improve the sim performance, its not a necessary metric for high quality DV.   

> 
> for debug and stress I think we already have good handle on these - maybe I don't fully understand what is meant by "planned".
> I can add going forward plans to the DV guide but then the state of the document does not match what is in the code. and I don't think it makes sense to create a new document stating what will go into the real document later!
> 

'Planned' simply means that tests corresponding to those categories (whichever applicable), with sufficient description indicating how will the said feature will be tested, have been entered in the testplan for review with wider audience. 

DV plan itself is not required to be fully complete at the V1 stage (and it doesn't have to match what's in the code), only thing you need to capture in there is a high level overview of what your testbench looks like and how you plan to go about verifying those features. This helps higher level testbench DV owners (such as myself) get a good sense of how to verify the IP at the chip level.   

> > * V2_CHECKLIST_SCOPED: This goes hand in hand with the previous item. Did you already start this @rasmus-madsen ?
> >   Yes this is scoped. in terms of man hours not actual time..
> 
> > The last two items are independent of the test plan review.
> > But all item are needed to start the actual V1 review.

To reiterate, V1 checklist review can be done in the PR, but the testplan and DV plan review would have to be done in a f2f meeting with a wider audience involving at least, the designer, chip lead, dv lead and 1-2 peers, higher level testbench owner (this would be the top earlgrey testbench), architect and SW. From DV perspective, we have 2 such 'high touch' reviews - this one and the one at the final sign off. I'm sure you might be following a similar approach for your internal projects at WD.

sriyerg commented on hw/ip/aes/doc/checklist.md:
```diff
@@ -103,41 +103,46 @@ Review        | Signoff date            | Not Started |
+Testbench     | [PRELIMINARY_ASSERTION_CHECKS_ADDED][]| Done        |
+Testbench     | [SIM_TB_ENV_CREATED][]                | Done        |
+Testbench     | [SIM_RAL_MODEL_GEN_AUTOMATED][]       | Done        |
+Testbench     | [CSR_CHECK_GEN_AUTOMATED][]           | ?           |
```
This is done, considering AES tb is derived from uvmtbgen script and uses the dvsim.py tool / make flow for running sims.

vogelpi replied:
Great, I marked this as done. Thanks for clarifying that @sriyerg .

rasmus-madsen commented:
> I was thinking more on the lines of blocking 30 mins of the next Si meeting (or setup an ad-hoc meeting) and presenting the testplan (what all things are being tested) and DV plan (how are those things tested) for a wider audience. It would be hard to do an "offline" review of the PR with all the stakeholders involved.


sounds good to me. Is there a process in place on who and how to arrange this?

sjgitty commented:
> I was thinking more on the lines of blocking 30 mins of the next Si meeting (or setup an ad-hoc meeting) and presenting the testplan (what all things are being tested) and DV plan (how are those things tested) for a wider audience.

I recommend not in the Si meeting, but in a one-off meeting for those that are required
and those that are interested.

vogelpi commented:
Thanks @sriyerg and @rasmus-madsen for clarifying the remaining points. I have updated this PR once more. Also, @rasmus-madsen has made a PR to update the DV plan, especially the item STD_TEST_CATEGORIES_PLANNED here: #1724  

I will send out an email for organize the meeting for the DV plan review. I would be great if we could have the meeting next week.

vogelpi commented:
I have rebased and updated once more based on the succesfull testplan review we had today.

sriyerg commented on hw/ip/aes/doc/checklist.md:
```diff
@@ -103,41 +103,46 @@ Review        | Signoff date            | Not Started |
+Tests         | [SIM_SANITY_TEST_PASSING][]           | Done        |
+Tests         | [SIM_CSR_MEM_TEST_SUITE_PASSING][]    | Done        |
+Tests         | [FPV_MAIN_ASSERTIONS_PROVEN][]        | N/A         |
+Tool Setup    | [SIM_ALT_TOOL_SETUP][]                | Done        |
```
```suggestion
Tool Setup    | [SIM_ALT_TOOL_SETUP][]                | Done        | Xcelium (signoff), VCS (alt)
```

vogelpi replied:
Done.

vogelpi commented:
Thanks everybody for the good discussion during the testplan review meeting yesterday, I am happy we can finally push this over the line.
